### ISSUE TITLE
HAVING in drift file causes null check error

### ISSUE THREAD

**Issue by austinried** [simolus3/moor#2378]
I'm not sure if this is actually just not supported currently or not, but I can't seem to get `HAVING` to work in my drift file.  I would really like to use it with a dart template variable so that I can provide the filters dynamically:
```sql
filterAlbums WITH Album:
  SELECT 
    albums.*,
    COUNT(songs.id) AS songs_count,
    SUM(CASE WHEN songs.download_file_path IS NOT NULL THEN 1 ELSE 0 END) AS downloaded_songs_count
  FROM albums
  LEFT JOIN songs ON albums.source_id = songs.source_id AND albums.id = songs.album_id
  WHERE $predicate
  GROUP BY albums.source_id, albums.id
  HAVING $having
  ORDER BY $order
  LIMIT $limit;
```
But this causes the builder to throw an error:
```
[SEVERE] drift_dev on lib/database/database.dart:

Null check operator used on a null value
```

If I change it to something static, like `HAVING songs_count > 0`, I get another error:
```
[WARNING] drift_dev on lib/database/database.dart:
line 340, column 10: Unknown column.
    ╷
340 │   HAVING songs_count > 0
    │          ^^^^^^^^^^^
    ╵
```

The only way it seems to work is if I include the entire aggregate in the expression: `HAVING COUNT(songs.id) > 0`.

**Comment by simolus3** [comment]
Thanks for the report! Both issues are bugs in drift, your code should have worked ;)

The "null check operator used on a null value" occurred because drift was unable to infer a type for `$having`. That was just a case I've missed, it should be `bool` of course - fixed in 834cc154bc654e4ced26e90406a33e293a55e3b5.

> `340 │   HAVING songs_count > 0`

Huh, I didn't know that was possible. I thought only `ORDER BY` expressions could refer to result columns like that, but `GROUP BY` and `HAVING` clauses can actually do that as well. Fixed in bdac301866c2ec9473e1898e33026aa9f1413685.

**Comment by austinried** [comment]
Awesome, thanks for such a quick response!
